File: practice/algorithms/sorting/insertion_sort-part_1/insertion_sort1.rb
def print_array(arr)
  puts arr.join(" ")
end

def insertionSort(ar)
  num = ar[ar.size - 1]
  placed = false
  (ar.size - 2).downto(0) do |j|
    if(ar[j] > num)
      ar[j + 1] = ar[j]
      print_array(ar)
    else
      ar[j + 1] = num
      print_array(ar)
      placed = true
      break
    end
  end
  if(!placed)
    ar[0] = num
    print_array(ar)
  end
end

count = gets.to_i
ar = gets.strip.split.map {|i| i.to_i}

insertionSort(ar)

File: practice/algorithms/sorting/insertion_sort-part_1/InsertionSort1.js
'use strict';

let insertIntoSorted = arr => {
    let num = arr[arr.length - 1];
    let placed = false;
    for(let j = arr.length - 2; j >= 0; j--) {
        if(arr[j] > num) {
            arr[j + 1] = arr[j];
            printArray(arr);
        } else {
            arr[j + 1] = num;
            printArray(arr);
            placed = true;
            break;
        }
    }
    if(!placed) {
        arr[0] = num;
        printArray(arr);
    }
};

let printArray = arr => console.log(arr.join(" "));

const processData = input => {
    let lines = input.split('\n');
    let arr = lines[1].split(' ').map(i => parseInt(i));
    insertIntoSorted(arr)
};

process.stdin.resume();
process.stdin.setEncoding("ascii");

var _input = "";
process.stdin.on("data", input => _input += input);
process.stdin.on("end", () => processData(_input));